packages/big-design/src/components/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,153 @@
+
+const excludeIconProps = <T extends any>({
+  iconOnly,
+  iconRight,
+  iconLeft,
+  ...actionProps
+}: T): Pick<T, Exclude<keyof T, 'iconOnly' | 'iconRight' | 'iconLeft'>> => actionProps;
+
+export const ButtonGroup: React.FC<ButtonGroupProps> = memo(({ actions, ...wrapperProps }) => {
+  const parentRef = createRef<HTMLDivElement>();
+  const dropdownRef = createRef<HTMLDivElement>();
+  const [isMenuVisible, setIsMenuVisible] = useState(false);
+  const [actionsState, setActionsState] = useState(
+    actions.map((action) => ({ isVisible: true, action: excludeIconProps(action), ref: createRef<HTMLDivElement>() })),
+  );
+
+  const changeActionsVisibility = useCallback(() => {
+    const parentWidth = parentRef.current?.offsetWidth;
+    const dropdownWidth = dropdownRef.current?.offsetWidth;
+
+    if (!parentWidth || !dropdownWidth) {
+      return;
+    }
+
+    let remainingWidth = parentWidth;
+
+    const nextState = actionsState.map((stateObj) => {
+      const actionWidth = stateObj.ref.current?.offsetWidth;
+
+      if (!actionWidth) {
+        return stateObj;
+      }
+
+      if (stateObj.action.actionType === 'destructive') {
+        return { ...stateObj, isVisible: false };
+      }
+
+      if (remainingWidth - actionWidth > dropdownWidth) {
+        remainingWidth = remainingWidth - actionWidth;
+
+        return { ...stateObj, isVisible: true };
+      }
+
+      return { ...stateObj, isVisible: false };
+    });
+
+    const visibleActions = actionsState.filter(({ isVisible }) => isVisible);
+    const nextVisibleActions = nextState.filter(({ isVisible }) => isVisible);
+
+    if (visibleActions.length !== nextVisibleActions.length) {
+      setActionsState(nextState);
+    }
+  }, [actionsState, dropdownRef, parentRef]);
+
+  const renderedDropdown = useMemo(
+    () => (
+      <StyledFlexItem data-testid="button-group-dropdown" isVisible={isMenuVisible} ref={dropdownRef} role="listitem">
+        <Dropdown
+          items={actionsState
+            .filter(({ isVisible }) => !isVisible)
+            .map(({ action, ref }) => ({
+              actionType: action.actionType,
+              content: action.text,
+              disabled: action.disabled,
+              onItemClick: () => {
+                if (ref.current) {
+                  ref.current.getElementsByTagName('button')[0].click();
+                }
+              },
+              hash: action.text.toLowerCase(),
+              icon: action.icon,
+            }))}
+          toggle={
+            <StyledButton
+              borderRadius={actionsState.every(({ isVisible }) => !isVisible)}
+              iconOnly={<MoreHorizIcon title="more" />}
+              variant="secondary"
+            />
+          }
+        />
+      </StyledFlexItem>
+    ),
+    [actionsState, dropdownRef, isMenuVisible],
+  );
+
+  const renderedActions = useMemo(
+    () =>
+      [...actionsState]
+        .reverse()
+        .sort(({ isVisible }) => (isVisible ? -1 : 1))
```

Comment:
Here I'm sorting elements that visible elements will be first in the row. I do this for the correct working of pseudo-class `:first-of-type` which controls `border-radius` for the left side of the first element.

it's the possible case when the first element will be hidden under ellispsis when this element will have `actionType="destructive"`.